### ISSUE TITLE
Process Hash Removes Entire Hash From URL Instead Of Just MSAL Properties

### ISSUE THREAD

**Issue by mindingdata** [AzureAD/microsoft-authentication-library-for-js#1422]
- [ ] `msal@1.x.x` or `@azure/msal@1.x.x`
- [x] `@azure/msal-browser@2.x.x`
- [ ] `@azure/msal-angular@0.x.x`
- [ ] `@azure/msal-angular@1.x.x`
- [ ] `@azure/msal-angularjs@1.x.x`

## Description

I have a particular scenario where I am using a Redirect Login, after login the user is redirected to an intermediary page where I check if we are in an IFrame (To stop the infinite loop), and if I am not, I then redirect back to my app using a URL I store in localstorage to redirect the user back to where they started with the id_token appended on the end. 

I set the `navigateToLoginRequestUrl` to false so this all works fine. 

However if I have a situation where a user may be emailed a link with an existing fragment/hash in the URL that jumps them to a specific spot on the page. For example : 

www.mysite.com/mypage#line103

If a user is sent this and they are not authenticated, I first store which page was requested in their local storage then push them to login. They will be redirected to my intermediary page after logging in looking like so : 

www.mysite.com/loginredirect.html#id_token=abc123

I then push the user back to the site they requested like so : 

www.mysite.com/mypage#line103&id_token=abc123

This logs them in fine but the main issue is that this line here : https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/37913c9129caca2e71d05a85f100263c92454c53/lib/msal-core/src/UserAgentApplication.ts#L1053

It removes various hash components of MSAL, but also deletes any other hash components. In my case not such a massive deal since it's only an auto scroll but I assume others who do more with Hash Routing may have larger issues. 

Is there a way in the future to use enum of ServerHashParamKeys and remove only the hash params that belong to MSAL and not those belonging to the app?

## Expected behavior
Would be great to not have MSAL remove hash components not belonging to the library.

**Comment by mvrak** [comment]
Hello Mindingdata,

We had direction to remove hash information as the reply urls do not support hash data.  So you are describing the intended behavior.  I'm not sure we would know what is strippable and what is not strippable.

There is a very similar concern among some app developers that reply URLs do not provide better support for wildcards.  That direction all comes from our security review.

Do you have any suggestions for how to tell what "#..." information should be allowed in an URL?

Thanks

**Comment by mindingdata** [comment]
Hi @mvrak, 

I think the best way would be to only strip the hash data MSAL knows about,  you already have an enum defined : 
https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/ea5a80f1293157e4105d0f4da6d89c3aea172860/lib/msal-core/src/utils/Constants.ts#L60

```
export enum ServerHashParamKeys {
    SCOPE = "scope",
    ERROR = "error",
    ERROR_DESCRIPTION = "error_description",
    ACCESS_TOKEN = "access_token",
    ID_TOKEN = "id_token",
    EXPIRES_IN = "expires_in",
    SESSION_STATE = "session_state",
    CLIENT_INFO = "client_info"
};
```

Remove all properties that match those, ignore the rest.

**Comment by jmckennon** [comment]
@mindingdata Can you clarify: which pages are running msal, and which pages are running your own custom code?

**Comment by jmckennon** [comment]
If you have any custom hashes, make sure that msal is NOT running on the same page--this may be a workaround in the interim.

**Comment by mindingdata** [comment]
>Can you clarify: which pages are running msal, and which pages are running your own custom code?

All pages are running MSAL because all pages need to be authenticated. Pretty standard scenario. 

The hash is only processed when returning from the login, but let's say that they are not currently logged in and user sends them a link to a page that has a hash in it, they will be forced to log in (Because it's an authenticated page), and then be returned to that page but then the hash is removed.

**Comment by tnorling** [comment]
@mindingdata A potential fix for this was included in ```msal@1.3.0-beta.1``` Can you try upgrading and let us know if this issue has been resolved?

**Comment by tnorling** [comment]
`msal@1.3.0` has been released and should fix this issue. Let us know if you're still having issues.
